### PR TITLE
Login logic pt.1

### PR DESCRIPTION
**Pull request by IvanKulundzic**
Added:
- network service
- login service
- keychain service

### CODE REVIEW

File: QuizApp/App/Services/LoginService.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import Foundation
+
+protocol LoginServiceProtocol {
+
+    func loginUser(_ password: String, username: String) async throws -> LoginResponseModel
```
Comment (istaudacher) [inline QuizApp/App/Services/LoginService.swift]: I think that just `login(username: String, password: String)` has enough info.

File: QuizApp/App/Screens/Login/LoginViewController.swift [hunk 2]
```diff
@@ -108,4 +126,21 @@ extension LoginViewController: ConstructViewsProtocol {
         }
     }
 
+    func addCallbacks() {
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewController.swift]: bindViewModel

File: QuizApp/App/Screens/Login/LoginViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    init(loginUseCase: LoginUseCaseProtocol) {
+        self.loginUseCase = loginUseCase
+    }
+
+}
+
+extension LoginViewModel {
+
+    func loginUser(password: String, username: String) async {
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewModel.swift]: Do not propagate async outside VM, handle it here.

File: QuizApp/App/Screens/Login/LoginViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        do {
+            let token = try await loginUseCase.loginUser(password, username: username).accessToken
+            /// Save the token to Keychain
+            let keychainService = SecureStorage()
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewModel.swift]: SecureStorage() can probably be a singleton. Also, UC or DS should handle saving the token, not VM.

File: QuizApp/App/Screens/Login/LoginViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            keychainService.save(token)
+        } catch {
+            handleError(error)
+        }
+    }
+
+}
+
+private extension LoginViewModel {
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewModel.swift:29]: Do you need extension?

File: QuizApp/App/Services/SecureStorage.swift [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import KeychainAccess
+
+protocol SecureStorageProtocol {
+    func save(_ token: String)
```
Comment (istaudacher) [inline QuizApp/App/Services/SecureStorage.swift:6]: newline before & after

File: QuizApp/App/Services/SecureStorage.swift [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+final class SecureStorage: SecureStorageProtocol {
+
+    private let keychainService = Keychain(service: "com.ivankulundzic.QuizApp")
+
+    func save(_ token: String) {
+        keychainService["ivan"] = token
```
Comment (istaudacher) [inline QuizApp/App/Services/SecureStorage.swift]: Please use a meaningful key.

File: QuizApp/App/Networking/NetworkClient.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import Foundation
+
+protocol NetworkClienProtocol {
```
Comment (istaudacher) [inline QuizApp/App/Networking/NetworkClient.swift]: `NetworkClientProtocol `

File: QuizApp/App/UseCase/LoginUseCase.swift [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+final class LoginUseCase: LoginUseCaseProtocol {
+
+    private let networkClient: NetworkClienProtocol
+
+    init(networkClient: NetworkClienProtocol) {
+        self.networkClient = networkClient
+    }
+
+    func loginUser(_ password: String, username: String) async throws -> LoginResponseModel {
+        guard let url = URL(string: Endpoint(type: .login).path) else {
```
Comment (istaudacher) [inline QuizApp/App/UseCase/LoginUseCase.swift]: Building URL requests should be the job of LoginNetworkClient.
Also, please add DataSource layer in between UC and Network.

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/App/Screens/Login/LoginViewModel.swift (hunk 5 above)
Comment (IvanKulundzic) [reply QuizApp/App/Screens/Login/LoginViewModel.swift:29]: Just a style I prefer. Even tho it is a simple app, but I prefer to group private methods in a private extension. Though I see value in identifying a method as private at a first glance if it has a private keyword.

File: QuizApp/App/Networking/Login/LoginDataSource.swift [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+protocol LoginDataSourceProtocol {
+
+    func login(_ username: String, password: String) async throws -> LoginDataModel
```
Comment (istaudacher) [inline QuizApp/App/Networking/Login/LoginDataSource.swift]: `func login(username: String, password: String) async throws -> LoginDataModel`

File: QuizApp/App/Networking/Login/LoginDataSource.swift [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+final class LoginDataSource: LoginDataSourceProtocol {
+
+    private let loginNetworkClient: LoginNetworkClientProtocol
+
+    init(loginNetworkClient: LoginNetworkClientProtocol) {
+        self.loginNetworkClient = loginNetworkClient
+    }
+
+    func login(_ username: String, password: String) async throws -> LoginDataModel {
+        guard let response = try? await loginNetworkClient.login(username, password: password) else {
```
Comment (istaudacher) [inline QuizApp/App/Networking/Login/LoginDataSource.swift]: You do not need to guard, just propagate error from network client

File: QuizApp/App/Screens/Login/LoginViewController.swift [hunk 12]
```diff
@@ -1,18 +1,35 @@
...
 final class LoginViewController: UIViewController {
 
     private var titleLabel: UILabel!
     private var emailInputField: InputField!
     private var passwordInputField: InputField!
     private var loginButton: UIButton!
+    private var cancellables = Set<AnyCancellable>()
+    private let loginViewModel: LoginViewModel
+
+    // MARK: - Init
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewController.swift]: please remove this comment

File: QuizApp/App/Screens/Login/LoginViewController.swift [hunk 13]
```diff
@@ -108,4 +126,21 @@ extension LoginViewController: ConstructViewsProtocol {
         }
     }
 
+    func bindViewModel() {
+        loginViewModel
+            .$errorLoggingIn
+            .sink { error in
+                print("Error logging in: \(error)")
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewController.swift]: Can you create an error label and present this error message?

File: QuizApp/App/Screens/Login/LoginViewController.swift [hunk 14]
```diff
@@ -108,4 +126,21 @@ extension LoginViewController: ConstructViewsProtocol {
...
+            }
+            .store(in: &cancellables)
+    }
+
+    @objc func loginButtonTapped() {
+        Task(priority: .background) {
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewController.swift]: Do you need to wrap this into a task? There aren't any async funcs.

File: QuizApp/App/Screens/Login/LoginViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,33 @@
+import Foundation
+import Combine
+
+final class LoginViewModel {
+
+    @Published var errorLoggingIn = ""
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewModel.swift]: @Published var errorMessage = ""

File: QuizApp/App/Screens/Login/LoginViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        self.loginUseCase = loginUseCase
+    }
+
+}
+
+extension LoginViewModel {
+
+    func loginUser(password: String, username: String) {
+        Task(priority: .background) {
+            await loginUseCase.login(username, password: password)
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewModel.swift]: Please handle error here.

File: QuizApp/App/Services/SecureStorage.swift [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import Foundation
+import KeychainAccess
+
+final class SecureStorage {
+
+    static let shared = SecureStorage()
```
Comment (istaudacher) [inline QuizApp/App/Services/SecureStorage.swift]: We avoid using static variables - they are not testable.

File: QuizApp/App/UseCase/LoginUseCase.swift [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+final class LoginUseCase: LoginUseCaseProtocol {
+
+    private let loginDataSource: LoginDataSourceProtocol
+
+    init(loginDataSource: LoginDataSourceProtocol) {
+        self.loginDataSource = loginDataSource
+    }
+
+    func login(_ username: String, password: String) async {
+        guard let token = try? await loginDataSource.login(username, password: password).accessToken else {
```
Comment (istaudacher) [inline QuizApp/App/UseCase/LoginUseCase.swift]: Propagate error.

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/App/Screens/Login/LoginViewController.swift (hunk 14 above)
Comment (IvanKulundzic) [reply QuizApp/App/Screens/Login/LoginViewController.swift]: Corrected. Didn't change it when I removed async from viewModel method.

File: QuizApp/App/Services/SecureStorage.swift (hunk 17 above)
Comment (IvanKulundzic) [reply QuizApp/App/Services/SecureStorage.swift]: The idea was to create a singleton for `SecureStorage`. What way would you recommend for creating one?

File: QuizApp/App/Screens/Login/LoginViewController.swift [hunk 19]
```diff
@@ -100,12 +125,40 @@ extension LoginViewController: ConstructViewsProtocol {
...
+    }
+
+    func bindViewModel() {
+        loginViewModel
+            .$errorMessage
+            .sink { [weak self] error in
+                guard let self = self else { return }
+
+                DispatchQueue.main.async {
+                    self.errorLabel.isHidden = false
```
Comment (istaudacher) [inline QuizApp/App/Screens/Login/LoginViewController.swift:149]: you need to guard self in async also ;) or guard it **just** in async

File: QuizApp/App/Services/SecureStorage.swift (hunk 17 above)
Comment (istaudacher) [reply QuizApp/App/Services/SecureStorage.swift]: Use that ServiceFactory and inject it where it is needed

**Review by istaudacher — CHANGES_REQUESTED** [review]

**Review by istaudacher — APPROVED** [review]

**Pull request merged**